File: sdk/go/harness/harness_test.go
package harness

import (
	"strings"
	"testing"

	"github.com/axisrobo/tekmovela-open/sdk/go/contracts"
)

func TestGoldenHarnessVersionDigest(t *testing.T) {
	hv, err := NewHarnessVersion(
		"checkout.double_spend",
		"v1",
		contracts.Reference{Kind: "VerificationContract", ID: "checkout.atomicity", Version: "v1", Digest: "sha256:" + repeat('a', 64)},
		sampleLock(),
		"tekmovela.evidence-bundle.v1.schema.json",
		DeterminismDeterministic,
		"",
	)
	if err != nil {
		t.Fatal(err)
	}
	d, err := hv.CanonicalDigest()
	if err != nil {
		t.Fatal(err)
	}
	if d != "sha256:bc8520405b1d90daea66d17a62c415f0edee8ecb4a52b9ed1460bc6a5a9b0499" {
		t.Fatalf("harness version digest %q != golden", d)
	}
}

func TestGoldenHarnessVersionDigestWithSeed(t *testing.T) {
	hv, err := NewHarnessVersion(
		"checkout.double_spend",
		"v1",
		contracts.Reference{Kind: "VerificationContract", ID: "checkout.atomicity", Version: "v1", Digest: "sha256:" + repeat('a', 64)},
		sampleLock(),
		"tekmovela.evidence-bundle.v1.schema.json",
		DeterminismDeterministic,
		"s1",
	)
	if err != nil {
		t.Fatal(err)
	}
	ec, ok := hv.Body()["environment_contract"]
	if !ok {
		t.Fatal("seed must emit environment_contract")
	}
	if got := ec.(map[string]any)["seed"]; got != "s1" {
		t.Fatalf("environment_contract.seed = %v, want s1", got)
	}
	d, err := hv.CanonicalDigest()
	if err != nil {
		t.Fatal(err)
	}
	if d != "sha256:ce4ba19c8f426de72786ca7b4206af833f487ee557e3444a056649023f110cd1" {
		t.Fatalf("seed harness version digest %q != golden", d)
	}
}

func TestGoldenHarnessVersionDigestStatistical(t *testing.T) {
	hv, err := NewHarnessVersion(
		"checkout.double_spend",
		"v1",
		contracts.Reference{Kind: "VerificationContract", ID: "checkout.atomicity", Version: "v1", Digest: "sha256:" + repeat('a', 64)},
		sampleLock(),
		"tekmovela.evidence-bundle.v1.schema.json",
		DeterminismStatistical,
		"",
	)
	if err != nil {
		t.Fatal(err)
	}
	d, err := hv.CanonicalDigest()
	if err != nil {
		t.Fatal(err)
	}
	if d != "sha256:ffc4b2e076bbb7588ac2923fcb6dc3f623016f022ab78daf0cc561e87046a519" {
		t.Fatalf("statistical harness version digest %q != golden", d)
	}
}

func TestPublishBindsDigest(t *testing.T) {
	hv, err := NewHarnessVersion(
		"checkout.double_spend",
		"v1",
		contracts.Reference{Kind: "VerificationContract", ID: "checkout.atomicity", Version: "v1", Digest: "sha256:" + repeat('a', 64)},
		sampleLock(),
		"tekmovela.evidence-bundle.v1.schema.json",
		DeterminismDeterministic,
		"",
	)
	if err != nil {
		t.Fatal(err)
	}
	want, err := hv.CanonicalDigest()
	if err != nil {
		t.Fatal(err)
	}
	if err := hv.Publish(); err != nil {
		t.Fatal(err)
	}
	if hv.Digest != want {
		t.Fatalf("Publish digest %q != canonical %q", hv.Digest, want)
	}
}

func TestNewHarnessVersionRejectsInvalidDeterminismMode(t *testing.T) {
	_, err := NewHarnessVersion(
		"checkout.double_spend",
		"v1",
		contracts.Reference{Kind: "VerificationContract", ID: "checkout.atomicity", Version: "v1", Digest: "sha256:" + repeat('a', 64)},
		sampleLock(),
		"tekmovela.evidence-bundle.v1.schema.json",
		DeterminismMode("random"),
		"",
	)
	if err == nil {
		t.Fatal("invalid determinism mode must fail")
	}
}

func TestHarnessVersionRejectsMissingComponent(t *testing.T) {
	_, err := NewHarnessVersion(
		"checkout.double_spend",
		"v1",
		contracts.Reference{Kind: "VerificationContract", ID: "checkout.atomicity", Version: "v1", Digest: "sha256:" + repeat('c', 64)},
		ComponentLock{
			ComponentRunner: {Kind: "HarnessComponent", ID: "r.1", Version: "v1", Digest: "sha256:" + repeat('c', 64)},
		},
		"evidence.bundle@v1",
		DeterminismDeterministic,
		"",
	)
	if err == nil {
		t.Fatal("missing component lock entries must fail")
	}
}

func TestLocalRunnerSealsBundle(t *testing.T) {
	runner := NewLocalRunner(&stubRuntime{})
	result, err := runner.Execute(
		Scenario{ID: "scenario.double_spend", Version: "v1", VerificationContractRef: "checkout.atomicity", Seed: "s1"},
		"cap/v2",
		"run.e0001",
	)
	if err != nil {
		t.Fatal(err)
	}
	if result.Bundle.BundleDigest == "" {
		t.Fatal("run bundle must be sealed")
	}
	if err := result.Bundle.Verify(); err != nil {
		t.Fatal(err)
	}
	if result.RunID != result.Bundle.RunRef {
		t.Fatalf("run id %q != bundle run ref %q", result.RunID, result.Bundle.RunRef)
	}
	if result.Status != string(RunStatusPassed) {
		t.Fatalf("run status %q != passed", result.Status)
	}
}

func TestExecuteDefaultsHarnessVersionRef(t *testing.T) {
	runner := NewLocalRunner(&stubRuntime{})
	result, err := runner.Execute(
		Scenario{ID: "scenario.double_spend", Version: "v1", VerificationContractRef: "checkout.atomicity", Seed: "s1"},
		"cap/v2",
		"run.e0002",
	)
	if err != nil {
		t.Fatal(err)
	}
	if result.Bundle.HarnessVersionRef != "local@v1" {
		t.Fatalf("harness version ref %q != local@v1", result.Bundle.HarnessVersionRef)
	}
}

func TestExecuteRecordsFailureStatus(t *testing.T) {
	runner := NewLocalRunner(&stubRuntime{status: RunStatusFailed})
	result, err := runner.Execute(
		Scenario{ID: "scenario.double_spend", Version: "v1", VerificationContractRef: "checkout.atomicity", Seed: "s1"},
		"cap/v2",
		"run.e0003",
	)
	if err != nil {
		t.Fatal(err)
	}
	if result.Status != string(RunStatusFailed) {
		t.Fatalf("run status %q != failed", result.Status)
	}
	if result.Bundle.BundleDigest == "" {
		t.Fatal("bundle must still be sealed on a failed run")
	}
	if err := result.Bundle.Verify(); err != nil {
		t.Fatal(err)
	}
}

type stubRuntime struct {
	status RunStatus
}

func (stubRuntime) StartRun(req RunRequest) (RunHandle, error) {
	return RunHandle{RunRef: req.ScenarioRef + "/run-" + req.Seed}, nil
}
func (stubRuntime) FetchEffects(handle RunHandle) ([]Effect, error) {
	return []Effect{{Action: "debit", ArgsDigest: "sha256:" + repeat('a', 64), Resource: "ledger", StateDelta: "-1"}}, nil
}
func (stubRuntime) CancelRun(RunHandle) error { return nil }
func (stubRuntime) CheckpointRun(RunHandle) (CheckpointResult, error) {
	return CheckpointResult{CheckpointRef: "ckp", EvidenceRef: "ev", Digest: "sha256:" + repeat('b', 64)}, nil
}
func (r stubRuntime) Status(RunHandle) (RunStatus, error) {
	if r.status == "" {
		return RunStatusPassed, nil
	}
	return r.status, nil
}
func (stubRuntime) Probe(ProbeRequest) (ProbeResult, error) {
	return ProbeResult{EvidenceRef: "ev/probe"}, nil
}

func sampleLock() ComponentLock {
	return ComponentLock{
		ComponentRunner:   {Kind: "HarnessComponent", ID: "runner.deterministic", Version: "v1", Digest: "sha256:" + repeat('c', 64)},
		ComponentAdapter:  {Kind: "HarnessComponent", ID: "adapter.praxovela", Version: "v1", Digest: "sha256:" + repeat('d', 64)},
		ComponentOracle:   {Kind: "HarnessComponent", ID: "oracle.effect_ledger", Version: "v1", Digest: "sha256:" + repeat('e', 64)},
		ComponentReporter: {Kind: "HarnessComponent", ID: "reporter.evidence", Version: "v1", Digest: "sha256:" + repeat('f', 64)},
	}
}

func repeat(c byte, n int) string {
	return strings.Repeat(string(c), n)
}

File: sdk/typescript/test/harness.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { HarnessVersion, LocalRunner } from "../src/harness.ts";
import type { RuntimeAdapter } from "../src/harness.ts";
import { ContractError, Reference } from "../src/contracts.ts";
import { ItemKind } from "../src/evidence.ts";

const D = "sha256:" + "a".repeat(64);
const VCONTRACT = new Reference("VerificationContract", "checkout.atomicity", "v1", D);

const GOLDEN_HARNESS_VERSION_DIGEST = "sha256:bc8520405b1d90daea66d17a62c415f0edee8ecb4a52b9ed1460bc6a5a9b0499";

function componentLock(): Record<string, { kind: string; id: string; version: string; digest: string }> {
  return {
    runner: { kind: "HarnessComponent", id: "runner.deterministic", version: "v1", digest: "sha256:" + "c".repeat(64) },
    adapter: { kind: "HarnessComponent", id: "adapter.praxovela", version: "v1", digest: "sha256:" + "d".repeat(64) },
    oracle: { kind: "HarnessComponent", id: "oracle.effect_ledger", version: "v1", digest: "sha256:" + "e".repeat(64) },
    reporter: { kind: "HarnessComponent", id: "reporter.evidence", version: "v1", digest: "sha256:" + "f".repeat(64) },
  };
}

function sampleHarnessVersion(seed?: string): HarnessVersion {
  return new HarnessVersion({
    id: "checkout.double_spend",
    version: "v1",
    verificationContractRef: VCONTRACT,
    componentLock: componentLock(),
    evidenceSchema: "tekmovela.evidence-bundle.v1.schema.json",
    ...(seed !== undefined ? { seed } : {}),
  });
}

function stubRuntime(): RuntimeAdapter {
  return {
    async startRun(req) {
      return { runRef: `${req.scenarioRef}/run-${req.seed}` };
    },
    async fetchEffects(handle) {
      return [{ action: "debit", argsDigest: "sha256:" + "a".repeat(64), resource: "ledger", stateDelta: "-1" }];
    },
    async cancelRun() {},
    async checkpointRun(handle) {
      return { checkpointRef: `ckp/${handle.runRef}`, evidenceRef: `evidence/${handle.runRef}`, digest: "sha256:" + "b".repeat(64) };
    },
    async status() {
      return "passed";
    },
    async probe() {
      return { effects: [], evidenceRef: "evidence/probe" };
    },
  };
}

test("harness version publishes a digest", () => {
  const hv = sampleHarnessVersion();
  hv.publish();
  assert.match(hv.digest ?? "", /^sha256:[0-9a-f]{64}$/);
});

test("harness version digest matches the shared cross-SDK golden digest", () => {
  const hv = sampleHarnessVersion();
  hv.publish();
  assert.equal(hv.digest, GOLDEN_HARNESS_VERSION_DIGEST);
});

test("harness version body emits the verification_contract_ref reference object", () => {
  const hv = sampleHarnessVersion();
  assert.deepEqual(hv.body().verification_contract_ref, {
    kind: "VerificationContract",
    id: "checkout.atomicity",
    version: "v1",
    digest: D,
  });
});

test("harness version body emits schema-required isolation_profile", () => {
  const hv = sampleHarnessVersion();
  assert.deepEqual(hv.body().isolation_profile, {});
});

test("harness version body omits environment_contract when seed is unset", () => {
  const hv = sampleHarnessVersion();
  assert.equal("environment_contract" in hv.body(), false);
});

test("harness version body emits environment_contract.seed when set", () => {
  const hv = sampleHarnessVersion("s1");
  assert.deepEqual(hv.body().environment_contract, { seed: "s1" });
});

test("harness version rejects a malformed verification_contract_ref", () => {
  assert.throws(
    () =>
      new HarnessVersion({
        id: "checkout.double_spend",
        version: "v1",
        verificationContractRef: { kind: "VerificationContract", id: "c", version: "v1", digest: "nope" },
        componentLock: componentLock(),
        evidenceSchema: "evidence.bundle@v1",
      }),
    ContractError
  );
});

test("harness version rejects a component lock entry without kind", () => {
  const lock = componentLock();
  delete lock.runner.kind;
  assert.throws(
    () =>
      new HarnessVersion({
        id: "checkout.double_spend",
        version: "v1",
        verificationContractRef: VCONTRACT,
        componentLock: lock,
        evidenceSchema: "evidence.bundle@v1",
      }),
    ContractError
  );
});

test("harness version rejects a component lock entry with a bad digest", () => {
  const lock = componentLock();
  lock.adapter.digest = "not-a-digest";
  assert.throws(
    () =>
      new HarnessVersion({
        id: "checkout.double_spend",
        version: "v1",
        verificationContractRef: VCONTRACT,
        componentLock: lock,
        evidenceSchema: "evidence.bundle@v1",
      }),
    ContractError
  );
});

test("harness version rejects an unsupported determinism mode", () => {
  assert.throws(
    () =>
      new HarnessVersion({
        id: "checkout.double_spend",
        version: "v1",
        verificationContractRef: VCONTRACT,
        componentLock: componentLock(),
        evidenceSchema: "evidence.bundle@v1",
        determinismMode: "random" as "statistical",
      }),
    ContractError
  );
});

test("local runner executes a scenario and seals a bundle", async () => {
  const runner = new LocalRunner(stubRuntime());
  const scenario = {
    id: "scenario.double_spend",
    version: "v1",
    verificationContractRef: "checkout.atomicity",
    seed: "s1",
    harnessVersionRef: "local@v2",
  };
  const result = await runner.execute(scenario, "cap/v2", "run.e0001");
  assert.ok(result.bundle.bundleDigest);
  result.bundle.verify();
  assert.ok(result.bundle.items.length > 0);
  assert.equal(result.runId, result.bundle.runRef);
  assert.equal(result.bundle.harnessVersionRef, scenario.harnessVersionRef);
  assert.equal(result.bundle.items[0].kind, ItemKind.Effect);
});

test("local runner defaults harness version to local@v1", async () => {
  const runner = new LocalRunner(stubRuntime());
  const result = await runner.execute(
    { id: "scenario.double_spend", version: "v1", verificationContractRef: "checkout.atomicity", seed: "s1" },
    "cap/v2",
    "run.e0002"
  );
  assert.equal(result.bundle.harnessVersionRef, "local@v1");
});
